### PR TITLE
Handle "window/showMessageRequest" in quick panel

### PR DESCRIPTION
**Pull request by ayoub-benali**
An attempt to have a proper fix for https://github.com/tomv564/LSP/issues/58 as "window/showMessageRequest" requires user action.

Example where such action would be needed: https://scalameta.org/metals/docs/editors/new-editor.html#import-build

### CODE REVIEW

File: plugin/core/rpc.py [hunk 1]
```diff
@@ -173,14 +176,15 @@ def on_notification(self, notification_method: str, handler: 'Callable') -> None
...
+        request_id = request.get("id")
         params = request.get("params", dict())
         method = request.get("method", '')
         debug('<--  ' + method)
         if self.settings.log_payloads and params:
             debug('     ' + str(params))
         if method in self._request_handlers:
             try:
-                self._request_handlers[method](params)
+                self._request_handlers[method](params, request_id)
```
Comment (rwols) [inline plugin/core/rpc.py]: Would adding this extra request_id argument not break the other request handlers? e.g. workspace/applyEdit?

File: plugin/core/protocol.py [hunk 2]
```diff
@@ -153,6 +153,20 @@ def to_payload(self, id) -> 'Dict[str, Any]':
         return r
 
 
+class Response:
+    def __init__(self, origin_request_id: str, result: 'Dict[str, Any]') -> None:
+        self.origin_request_id = origin_request_id
```
[the remark this answers is not in the document]
Comment (ayoub-benali) [reply plugin/core/protocol.py]: :+1:

File: plugin/core/windows.py [hunk 3]
```diff
@@ -428,7 +436,7 @@ def _handle_session_started(self, session, project_path, config):
 
         client.on_request(
             "window/showMessageRequest",
-            lambda params: self._handle_message_request(params))
+            lambda params, request_id: self._handle_message_request(params, client, request_id))
```
[the remark this answers is not in the document]
Comment (ayoub-benali) [reply plugin/core/windows.py]: yes will update the other handlers

**Comment by ayoub-benali** [comment]
@rwols I updated the code regarding your comments and Travis seems happy. can you please re-check?

**Comment by ayoub-benali** [comment]
@rwols I was also searching how to add a test for that but didn't find the entry point. If you can guide me I will add tests

**Comment by ayoub-benali** [comment]
> So I think (but I could be wrong) that the correct strategy is to write a test for window/showMessageRequest in tests/ and check that a quick panel appears.
We also have to mock the client to check that it is sending a message back with the user chose to the server, right ?

**Comment by rwols** [comment]
I think you can use a MockTransport class for that. It is located in plugin/core/test_rpc.py.

**Comment by ayoub-benali** [comment]
@rwols thanks ! When can you issue a new release for the plug-in ?

**Comment by rwols** [comment]
I will leave that up to @tomv564 :)

**Comment by tomv564** [comment]
I'll aim for a new release soon.
Thanks for this great feature, soon there will be no TODOs left in this project 😮 

Regarding the no-op comment in `send_user_choice`, perhaps we should respond anyways, with a `null` if nothing was selected? (reading https://microsoft.github.io/language-server-protocol/specification#window_showMessageRequest)

**Comment by ayoub-benali** [comment]
you are right we need to reply anyway. I will issue a PR to fix this. Thanks for pointing it out

**Pull request merged**